### ISSUE TITLE
F5 parsing errors

### ISSUE THREAD

**Issue by Ioda100** [mpenning/ciscoconfparse#145]
When i try to pare such F5 formatted files, it doesn't work.

1) 

> if {[active_members APN-DNS-TCP] > 0 & [active_members APN-DNS-UDP] > 0  } {

This kind of lines with opening and closing brackets on the same line seem to cause issues.

2) it seems that in F5 format there is also a kind of banner where we can put multilines text with:

>     include  "
> ...
> "

But the parser is not seeing it the same way than banner motd  in cisco ios

3) Even if we try to deal with include " ... "  as the banners, it fails before because  in the text there are some   };  that are seen as closing sections and not text like here:

> sys syslog {
>     include "
> template t_remotetmpl {
> template (\"<$PRI>$STAMP $HOST $FACILITY[$PID]: $MSGONLY\"); template_escape(no);
> };
> "

Extract of the F5 config file that doesn't work:

> ltm profile udp DNS-UDP {
>     app-service none
>     datagram-load-balancing disabled
>     idle-timeout 31
> }
> ltm rule contrail-monitor {
>     when HTTP_REQUEST {
>                 **if {[active_members APN-DNS-TCP] > 0 & [active_members APN-DNS-UDP] > 0  } {**
>                         HTTP::respond 200 content "up"
>                 }
>         }
> }
> ltm rule contrail-monitor1 {
>     when HTTP_REQUEST {
>                 **if {[active_members APN-DNS-TCP] >= 0 & [active_members APN-DNS-UDP] >= 0  }** {
>                         HTTP::respond 200 content "up"
>                 }
>         }
> }
> ltm tacdb licenseddb licensed-tacdb {
>     partition none
> }
> sys state-mirroring { }
> sys syslog {
>     **include "**
> template t_remotetmpl {
> template (\"<$PRI>$STAMP $HOST $FACILITY[$PID]: $MSGONLY\"); template_escape(no);
> **};**
> filter f_remote_loghost {
> level(info..emerg);
> **};**
> destination d_remote_loghost {
> udp(\"102.223.51.181\" port(519) template(t_remotetmpl));
> **};**
> log {
> source(s_syslog_pipe);
> filter(f_remote_loghost);
> destination(d_remote_loghost);
> **};**
> **"**
> remote-servers {
>         JSA {
>             host 102.223.51.181
>         }
>     }
> }
> sys url-db download-schedule urldb { }

**Comment by mpenning** [comment]
Hello @Ioda100  and thank you for this bug report.

Regarding point 1: I'm not sure I see what issue you're referring to... can you be more explicit about the problem you see?

Regarding point 2: I think you're suggesting that I include this pattern as a kind of comment block; however, at this time I'm leveraging the junos parser as the f5 parser as well, so I'm not ready to make a special case for this as a comment delimiter (since this could break junos configs).  Is this going to be a big problem for you?

Regarding point 3: I just pushed some changes to ciscoconfparse's git master that should fix the parsing of this line.  Please test with the latest commits (not yet pushed out as a formal ciscoconfparse version).

As a side note, I also created an example F5 config based on the config in this github issue, but I wound up deleting a bunch of characters like ** (double asterisks) because this looked like markdown formatting in the middle of your F5 config to me.  Can you do a diff against `configs/sample_01.f5` and ensure that this really is the same as your running config?

**Comment by Ioda100** [comment]
For point 1, i mean that sending this to the parser:

> ltm rule contrail-monitor1 {
>         when HTTP_REQUEST {
>                 if {[active_members APN-DNS-TCP] > 0 & [active_members APN-DNS-UDP] > 0  } {
>                         HTTP::respond 200 content "up"
>                 }
>         }
> }
> ltm rule contrail-monitor {
>         when HTTP_REQUEST {
>                 if {[active_members APN-DNS-TCP] > 0 & [active_members APN-DNS-UDP] > 0  } {
>                         HTTP::respond 200 content "up"
>                 }
>         }
> }

**when.HTTP_REQUEST**  of the second block is not seen as child of **contrail-monitor** and it seems to be because of the brackets on the same line:  **} {**


For point 2,  Not an issue for us, we can workaround. It was just to make you aware this seems to exist in F5 format.

For point 3, ok thanks i will test and let you know if not working :-)

The file looks good. Sorry for the ** . I put in bold the elements causing issue in the file and i think it puts ** to do it. 

Thanks for your great parser again.

**Comment by mpenning** [comment]
Hello,

Parsing like this: `CiscoConfParse('configs/sample_01.f5', syntax='junos')`:

```
ltm profile udp DNS-UDP
    app-service none
    datagram-load-balancing disabled
    idle-timeout 31
ltm rule contrail-monitor
    when HTTP_REQUEST
        if {[active_members APN-DNS-TCP] > 0 & [active_members APN-DNS-UDP] > 0  }
            HTTP::respond 200 content "up"
ltm rule contrail-monitor1
    when HTTP_REQUEST
        if {[active_members APN-DNS-TCP] >= 0 & [active_members APN-DNS-UDP] >= 0  }
            HTTP::respond 200 content "up"
```

I still don't see the same problem you describe... are you getting different results?  If so, do you have the latest version of CiscoConfParse?

**Comment by Ioda100** [comment]
Yes you're right for problem3 it seems that it's the pre-treatment done for the include " problem that leads to that error. Using the lib alone it's working fine.

**Comment by mpenning** [comment]
The official fix is already present in git master, but will also be published with version 1.3.38 (soon to be released)
